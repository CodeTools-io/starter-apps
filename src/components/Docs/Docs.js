import React, { useMemo, useState } from 'react';
import {
  Box,
  Card,
  Grid,
  Heading,
  List,
  Markdown,
  Paragraph,
  Tab,
  Tabs,
  Text,
  ThemeContext,
} from 'grommet';
// import MonacoEditor from 'react-monaco-editor';
import MonacoEditor from '@monaco-editor/react';
import theme from './theme';

const options = {
  fontSize: 16,
  minimap: {
    enabled: false,
  },
  readOnly: true,
};
export default function Docs({ children, ...props }) {
  return (
    <Box className="Docs" {...props}>
      {children}
    </Box>
  );
}

export function DocsTheme({ children, ...props }) {
  return (
    <ThemeContext.Extend value={theme} {...props}>
      {children}
    </ThemeContext.Extend>
  );
}

export function DocsCard({ children, ...props }) {
  return (
    <Card className="DocsCard" background="white" {...props}>
      <DocsTheme>{children}</DocsTheme>
    </Card>
  );
}
export function DocsOverview({ name, description }) {
  return (
    <Box gap="small" flex={false}>
      <Heading level={3} margin="none">
        {name}
      </Heading>
      <Paragraph margin="none" fill>
        {description}
      </Paragraph>
    </Box>
  );
}

export function DocsExample({ children, usage }) {
  const [active, setActive] = useState(0);
  const showPreview = useMemo(() => active === 0, [active]);
  const showCode = useMemo(() => active === 1, [active]);
  return (
    <Box gap="small" flex={false}>
      <Box direction="row" justify="between">
        <Heading level={4} margin="none"></Heading>
        <Tabs activeIndex={active} onActive={(next) => setActive(next)}>
          <Tab title="Preview"></Tab>
          <Tab title="Code"></Tab>
        </Tabs>
      </Box>
      {showPreview && (
        <DocsCard height="large" flex={false}>
          <Box overflow="auto">{children}</Box>
        </DocsCard>
      )}
      {showCode && (
        <DocsCard height="large" pad="medium" flex={false}>
          <MonacoEditor
            language="javascript"
            theme="light"
            height="100%"
            value={usage?.trim()}
            options={options}
          />
        </DocsCard>
      )}
    </Box>
  );
}

export function DocsProps({ properties }) {
  return (
    <Box gap="small" flex={false}>
      <Heading level={4} margin="none">
        Properties
      </Heading>
      <DocsCard>
        <List data={properties} pad="none">
          {(datum, index) => {
            const isAlternate = index % 2;
            const { name, description, format } = datum;
            const { overview, args, returnValue } = description;
            const formatAppend = args
              ? `(${args?.map((a) => a?.name)?.join(', ')})`
              : ``;
            const areas = [
              ['name', 'format', 'description'],
              args ? ['properties', 'properties', 'properties'] : null,
              returnValue
                ? ['returnValue', 'returnValue', 'returnValue']
                : null,
            ]?.filter((area) => area);
            return (
              <Box background={isAlternate ? 'light-1' : 'white'}>
                <Grid
                  columns={['auto', 'flex', '1/2']}
                  rows={['auto']}
                  areas={areas}
                  pad="medium"
                  gap={{ row: 'medium', column: 'small' }}
                  align="center"
                >
                  <Box gridArea="name" direction="row" gap="small">
                    <Text weight="bold">{name}</Text>
                  </Box>
                  <Box gridArea="format">
                    <Markdown>{`\`${format}${formatAppend}\``}</Markdown>
                  </Box>
                  <Paragraph
                    gridArea="description"
                    margin="none"
                    textAlign="end"
                    fill
                  >
                    {overview || description}
                  </Paragraph>
                  {args && (
                    <Box
                      gridArea="properties"
                      pad={{
                        vertical: 'small',
                        horizontal: 'medium',
                      }}
                      border
                    >
                      <Box gap="medium" border="between">
                        {args?.map((arg, index) => {
                          const {
                            name: argName,
                            type: argType,
                            description: argDescription,
                          } = arg;

                          return (
                            <Box
                              key={`${name}-arg-${argName}`}
                              direction="row"
                              align="baseline"
                              justify="between"
                              gap="small"
                            >
                              <Markdown>{`\`${argName} <${argType}>\``}</Markdown>
                              <Text color="dark-4">{argDescription}</Text>
                            </Box>
                          );
                        })}
                      </Box>
                    </Box>
                  )}
                  {returnValue && (
                    <Box
                      gridArea="returnValue"
                      direction="row"
                      align="baseline"
                      justify="start"
                      gap="small"
                    >
                      <Text color="dark-4">returns</Text>
                      <Markdown>{`\`${returnValue}\``}</Markdown>
                    </Box>
                  )}
                </Grid>
              </Box>
            );
          }}
        </List>
      </DocsCard>
    </Box>
  );
}

export function DocsPage({ component: Component, ...props }) {
  const { name, description, properties, usage } = Component.toJSON();

  return (
    <Box className="DocsPage" gap="large" fill="horizontal" {...props}>
      <DocsOverview name={name} description={description} />
      <DocsExample usage={usage}>
        <Component />
      </DocsExample>
      <DocsProps properties={properties} />
      <Box fill></Box>
    </Box>
  );
}
