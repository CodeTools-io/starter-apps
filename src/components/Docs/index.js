export { default, DocsCard, DocsTheme } from './Docs';
