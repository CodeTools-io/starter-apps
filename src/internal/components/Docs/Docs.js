import React, { useEffect, useMemo, useState } from 'react';
import {
  Box,
  Button,
  Card,
  Grid,
  Heading,
  Layer,
  List,
  Markdown,
  Paragraph,
  Tab,
  Tabs,
  Text,
  ThemeContext,
} from 'grommet';
import { CodeSandbox, Close, Expand } from 'grommet-icons';
import MonacoEditor from '@monaco-editor/react';
import Feature from 'internal/components/Feature';

import theme from './theme';

const DOCS_BASE_PATH = `${process.env.PUBLIC_URL}/docs`;
const SANDBOX_URL = process.env.REACT_APP_SANDBOX_URL;

const options = {
  fontSize: 16,
  minimap: {
    enabled: false,
  },
  readOnly: true,
};

export default function Docs({ children, ...props }) {
  return (
    <Box className="Docs" {...props}>
      {children}
    </Box>
  );
}

export function DocsCode({ files = [], options }) {
  const [activeFileIndex, setActiveFileIndex] = useState(null);

  useEffect(() => {
    if (activeFileIndex === null && files?.length) {
      setActiveFileIndex(
        files?.findIndex((file) => file?.filename === `${file?.app}.js`)
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [activeFileIndex, files]);

  return (
    <DocsCard height="large" pad="medium" flex={false}>
      <Box border fill>
        <Grid columns={['auto', 'flex']} rows={['full']} fill>
          <Box overflow={{ vertical: 'auto' }}>
            {files.map((file, index) => {
              return (
                <Box
                  key={`file-${file?.app}-${file?.filepath}`}
                  background={index === activeFileIndex ? 'white' : 'shade-1'}
                  border={[
                    'bottom',
                    {
                      side: 'right',
                      color:
                        index === activeFileIndex ? 'transparent' : 'border',
                    },
                  ]}
                  direction="row"
                  justify="center"
                  pad="small"
                  flex={false}
                >
                  <Button
                    label={file?.filepath}
                    onClick={() => setActiveFileIndex(index)}
                    size="large"
                    fill
                    plain
                  />
                </Box>
              );
            })}
            <Box border="right" flex />
          </Box>
          <Box>
            <MonacoEditor
              language="javascript"
              theme="light"
              height="100%"
              value={files[activeFileIndex]?.content}
              options={options}
            />
          </Box>
        </Grid>
      </Box>
    </DocsCard>
  );
}

export function DocsTheme({ children, ...props }) {
  return (
    <ThemeContext.Extend value={theme} {...props}>
      {children}
    </ThemeContext.Extend>
  );
}

export function DocsCard({ children, ...props }) {
  return (
    <Card className="DocsCard" background="white" {...props}>
      <DocsTheme>{children}</DocsTheme>
    </Card>
  );
}
export function DocsOverview({ name, description }) {
  return (
    <Box gap="small" flex={false}>
      <Heading level={3} margin="none">
        {name}
      </Heading>
      <Paragraph margin="none" fill>
        {description}
      </Paragraph>
    </Box>
  );
}

export function DocsPreviewModal({ children, onShrink = () => {} }) {
  return (
    <ThemeContext.Extend
      value={{
        layer: {
          background: 'transparent',
        },
      }}
    >
      <Layer
        onEsc={onShrink}
        onClickOutside={onShrink}
        margin={{
          horizontal: 'large',
          vertical: 'medium',
        }}
        full
        modal
      >
        <Box
          direction="row"
          justify="end"
          flex={false}
          background="transparent"
        >
          <Button icon={<Close color="white" />} onClick={() => onShrink()} />
        </Box>
        <Box height="100%" pad={{ top: 'medium' }}>
          <DocsPreview onShrink={onShrink} fill>
            {children}
          </DocsPreview>
        </Box>
      </Layer>
    </ThemeContext.Extend>
  );
}
export function DocsPreview({
  children,
  fullScreen = false,
  onShrink = () => {},
  ...props
}) {
  if (fullScreen) {
    return <DocsPreviewModal onShrink={onShrink}>{children}</DocsPreviewModal>;
  }

  return (
    <DocsCard height="large" flex={false} {...props}>
      <Box overflow="auto" fill>
        {children}
      </Box>
    </DocsCard>
  );
}

export function DocsMain({ children, files = [], sandboxUrl }) {
  const [active, setActive] = useState(0);
  const [fullScreen, setFullScreen] = useState(false);
  const showPreview = useMemo(() => active === 0, [active]);
  const showCode = useMemo(() => active === 1, [active]);

  return (
    <Box gap="small" flex={false}>
      <Box direction="row" justify="between">
        <Tabs activeIndex={active} onActive={(next) => setActive(next)}>
          <Tab title="Preview"></Tab>
          <Tab title="Code"></Tab>
        </Tabs>
        <Box direction="row" align="center" gap="small">
          <Button
            onClick={() => setFullScreen(true)}
            icon={<Expand size="small" />}
            color="control"
            plain
          />
          <Button
            icon={<CodeSandbox size="medium" />}
            color="control"
            href={sandboxUrl}
            plain
          />
        </Box>
      </Box>
      {showPreview && (
        <DocsPreview
          children={children}
          fullScreen={fullScreen}
          onShrink={() => setFullScreen(false)}
        />
      )}
      {showCode && <DocsCode files={files} options={options} />}
    </Box>
  );
}

export function DocsProps({ properties }) {
  return (
    <Box gap="small" flex={false}>
      <Heading level={4} margin="none">
        Properties
      </Heading>
      <DocsCard>
        <List data={properties} pad="none">
          {(datum, index) => {
            const isAlternate = index % 2;
            const { name, description, format } = datum;
            const { overview, args, returnValue } = description;
            const formatAppend = args
              ? `(${args?.map((a) => a?.name)?.join(', ')})`
              : ``;
            const areas = [
              ['name', 'format', 'description'],
              args ? ['properties', 'properties', 'properties'] : null,
              returnValue
                ? ['returnValue', 'returnValue', 'returnValue']
                : null,
            ]?.filter((area) => area);
            return (
              <Box background={isAlternate ? 'light-1' : 'white'}>
                <Grid
                  columns={['auto', 'flex', '1/2']}
                  rows={['auto']}
                  areas={areas}
                  pad="medium"
                  gap={{ row: 'medium', column: 'small' }}
                  align="center"
                >
                  <Box gridArea="name" direction="row" gap="small">
                    <Text weight="bold">{name}</Text>
                  </Box>
                  <Box gridArea="format">
                    <Markdown>{`\`${format}${formatAppend}\``}</Markdown>
                  </Box>
                  <Paragraph
                    gridArea="description"
                    margin="none"
                    textAlign="end"
                    fill
                  >
                    {overview || description}
                  </Paragraph>
                  {args && (
                    <Box
                      gridArea="properties"
                      pad={{
                        vertical: 'small',
                        horizontal: 'medium',
                      }}
                      border
                    >
                      <Box gap="medium" border="between">
                        {args?.map((arg, index) => {
                          const {
                            name: argName,
                            type: argType,
                            description: argDescription,
                          } = arg;

                          return (
                            <Box
                              key={`${name}-arg-${argName}`}
                              direction="row"
                              align="baseline"
                              justify="between"
                              gap="small"
                            >
                              <Markdown>{`\`${argName} <${argType}>\``}</Markdown>
                              <Text color="dark-4">{argDescription}</Text>
                            </Box>
                          );
                        })}
                      </Box>
                    </Box>
                  )}
                  {returnValue && (
                    <Box
                      gridArea="returnValue"
                      direction="row"
                      align="baseline"
                      justify="start"
                      gap="small"
                    >
                      <Text color="dark-4">returns</Text>
                      <Markdown>{`\`${returnValue}\``}</Markdown>
                    </Box>
                  )}
                </Grid>
              </Box>
            );
          }}
        </List>
      </DocsCard>
    </Box>
  );
}

export function DocsPage({ component: Component, ...props }) {
  const { name, description, properties } = Component.toJSON();
  const [data, setData] = useState();

  useEffect(() => {
    if (name) {
      fetch(`${DOCS_BASE_PATH}/${name}.json`)
        .then((res) => res.json())
        .then((data) => {
          setData(data);
        })
        .catch((err) => console.error(err));
    }
  }, [name]);

  return (
    <Box className="DocsPage" gap="large" fill="horizontal" {...props}>
      <DocsOverview name={name} description={description} />
      <DocsMain files={data?.files || []} sandboxUrl={`${SANDBOX_URL}${name}`}>
        <Component />
      </DocsMain>
      <Feature name="DOC_PROPS">
        <DocsProps properties={properties} />
      </Feature>
      <Box fill></Box>
    </Box>
  );
}
