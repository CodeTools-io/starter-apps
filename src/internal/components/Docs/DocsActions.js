import React from 'react';
import { Box } from 'grommet';

export default function DocsActions({ actions }) {
  if (!actions?.length) {
    return null;
  }

  return (
    <Box direction="row" align="center" gap="medium">
      {actions}
    </Box>
  );
}
