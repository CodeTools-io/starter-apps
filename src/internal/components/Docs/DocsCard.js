import React from 'react';
import { Card } from 'grommet';

export default function DocsCard({ children, ...props }) {
  return (
    <Card className="DocsCard" background="white" {...props}>
      {children}
    </Card>
  );
}
