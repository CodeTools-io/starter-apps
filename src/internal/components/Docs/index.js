export { default as Docs } from './Docs';
export { default as DocsCard } from './DocsCard';
export { default as DocsCode } from './DocsCode';
export { default as DocsMain } from './DocsMain';
export { default as DocsPreview } from './DocsPreview';
export { default as DocsPreviewModal } from './DocsPreviewModal';
export { default as DocsProps } from './DocsProps';
export { default as DocsComponents } from './DocsComponents';
