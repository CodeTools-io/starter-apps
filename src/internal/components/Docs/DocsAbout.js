import React, { useMemo, lazy } from 'react';
import { Box, Markdown, Heading, Paragraph } from 'grommet';

import DocsCard from './DocsCard';

function DocsAboutInlineCode({ children }) {
  return (
    <Box style={{ display: 'inline' }} pad={{ horizontal: 'xsmall' }} border>
      <code>{children}</code>
    </Box>
  );
}

function DocsAboutPreview({ children, caption = 'Preview', ...props }) {
  return (
    <>
      <Heading level="6" margin={{ bottom: 'xsmall' }}>
        {caption}
      </Heading>
      <Box
        style={{ position: 'relative' }}
        height={{ min: 'small' }}
        fill="horizontal"
        flex={false}
        border={{
          side: 'all',
          style: 'dashed',
        }}
        {...props}
      >
        {children}
      </Box>
    </>
  );
}
export default function DocsAbout({ doc }) {
  const Components = useMemo(() => {
    if (!doc?.components) {
      return {};
    }

    return Object.keys(doc?.components)?.reduce((accum, componentName) => {
      const Component = lazy(() =>
        import(`../../../${doc?.directory}/${componentName}`)
      );
      return {
        ...accum,
        [componentName]: {
          component: (props) => {
            const parsedProps = Object.entries(props).reduce(
              (accum, [key, value]) => {
                const isObject =
                  typeof value === 'string' &&
                  value.trim().startsWith('{{') &&
                  value.trim().endsWith('}}');
                const isArray =
                  typeof value === 'string' &&
                  value.trim().startsWith('{[') &&
                  value.trim().endsWith(']}');
                if (isObject) {
                  const rawValue = value.trim().slice(1, -1);
                  const newValue = JSON.parse(rawValue);
                  return {
                    ...accum,
                    [key]: newValue,
                  };
                }

                if (isArray) {
                  const rawValue = value.trim().slice(1, -1);
                  const newValue = JSON.parse(rawValue);

                  return {
                    ...accum,
                    [key]: newValue,
                  };
                }

                return {
                  ...accum,
                  [key]: value,
                };
              },
              {}
            );

            return <Component {...parsedProps} />;
          },
        },
      };
    }, {});
  }, [doc]);

  return (
    <Box className="DocsAbout" gap="small" flex={false}>
      <DocsCard>
        <Box pad={{ horizontal: 'large', top: 'small', bottom: 'large' }}>
          {doc?.content && (
            <Markdown
              components={{
                h1: {
                  component: Heading,
                  props: {
                    level: 1,
                    margin: { top: 'large', bottom: 'small' },
                  },
                },
                h2: {
                  component: Heading,
                  props: {
                    level: 2,
                    margin: { top: 'large', bottom: 'small' },
                  },
                },
                h3: {
                  component: Heading,
                  props: {
                    level: 3,
                    margin: { top: 'large', bottom: 'xsmall' },
                  },
                },
                h4: {
                  component: Heading,
                  props: {
                    level: 4,
                    margin: { top: 'medium', bottom: 'xsmall' },
                  },
                },
                h5: {
                  component: Heading,
                  props: {
                    level: 5,
                    margin: { top: 'medium', bottom: 'xsmall' },
                  },
                },
                h6: {
                  component: Heading,
                  props: {
                    level: 6,
                    margin: { top: 'medium', bottom: 'xsmall' },
                  },
                },
                p: {
                  component: Paragraph,
                  props: { size: 'medium', fill: true },
                },
                code: {
                  component: DocsAboutInlineCode,
                },
                hr: {
                  component: () => (
                    <Box
                      border={{ side: 'bottom' }}
                      margin={{ top: 'xlarge', bottom: 'xlarge' }}
                    />
                  ),
                },
                Preview: {
                  component: DocsAboutPreview,
                },
                ...Components,
              }}
            >
              {doc?.content}
            </Markdown>
          )}
        </Box>
      </DocsCard>
    </Box>
  );
}
