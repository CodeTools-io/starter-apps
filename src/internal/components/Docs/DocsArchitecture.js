import React, { useEffect, useState } from 'react';
import { Box } from 'grommet';

import DocsCard from './DocsCard';

function DocsArchitectureOverlay({ children, doc }) {
  const [styles, setStyles] = useState('');
  useEffect(() => {
    if (doc?.components) {
      const rootStyles = ``;
      const updatedStyles = Object.entries(doc?.components)
        .map(([componentName, description]) => {
          if (componentName === doc?.id) {
            return ``;
          }

          return `
          .DocsArchitectureOverlay .${componentName} {
            position: relative;
            box-sizing: border-box;
            pointer-events: none;
          }

          .DocsArchitectureOverlay .${componentName}:before {
            content: '';
            box-sizing: border-box;
            position: absolute;
            display: block;
            height: 100%;
            width: 100%;
            top: 0;
            left: 0;
            background: rgba(0,0,0,0.5);
            z-index: 900000;
          }
          .DocsArchitectureOverlay .${componentName}:after {
            content: '${componentName}';
            box-sizing: border-box;
            position: absolute;
            height: 100%;
            width: 100%;
            top: 0;
            left: 0;
            display: flex;
            border: 5px dashed #fff;
            justify-content: center;
            align-items: center;
            color: #fff;
            font-weight: bold;
            z-index: 900001;
          }
        `;
        })
        .join('\n');
      setStyles(`
      ${rootStyles}
      ${updatedStyles}
      `);
    }
  }, [doc]);

  return (
    <Box
      className="DocsArchitectureOverlay"
      style={{
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
      }}
    >
      <style type="text/css">{styles}</style>
      {children}
    </Box>
  );
}
export default function DocsArchitecture({ children, doc, ...props }) {
  return (
    <DocsCard height="large" flex={false} {...props}>
      <Box overflow="auto" style={{ position: 'relative' }} fill>
        {children}
        <DocsArchitectureOverlay doc={doc}>{children}</DocsArchitectureOverlay>
      </Box>
    </DocsCard>
  );
}
