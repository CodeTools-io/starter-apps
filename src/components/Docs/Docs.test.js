import React from 'react'
import Docs from './Docs'

describe('Docs', () => {
  test('can render', () => {
    const { container } = renderComponent(
      <Docs>
        <p>Content for the Docs</p>
      </Docs>
    )

    expect(container).toMatchSnapshot()
  })
})

