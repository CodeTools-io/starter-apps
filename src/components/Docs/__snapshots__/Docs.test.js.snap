// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Docs can render 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 bFwQzJ Docs"
  >
    <p>
      Content for the Docs
    </p>
  </div>
</div>
`;
